stellargraph/connector/neo4j/sampler.py

```diff
@@ -0,0 +1,202 @@
+    It can be used to extract a random sub-graph starting from a set of initial nodes from Neo4J database.
+    """
+
+    def run(
+        self, neo4j_graphdb, nodes=None, n=1, in_size=None, out_size=None, seed=None
+    ):
+        """
+        Send queries to Neo4J databases and collect sampled breadth-first walks starting from the root nodes.
+
+        Args:
+            neo4j_graphdb (py2neo.Graph): the Neo4J Graph Database object
+            nodes (list): A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n (int): Number of walks per node id.
+            in_size (list): The number of in-directed nodes to sample with replacement at each depth of the walk.
+            out_size (list): The number of out-directed nodes to sample with replacement at each depth of the walk.
+            seed (int): Random number generator seed; default is None
+        Returns:
+            A list of multi-hop neighbourhood samples. Each sample expresses multiple undirected walks, but the in-node
+            neighbours and out-node neighbours are sampled separately. Each sample has the format:
+        """
+
+        def bfs_neighbor_query(sampling_direction):
+            direction = {"BOTH": "--", "IN": "<--", "OUT": "-->"}
+
+            return f"""
+                // expand the list of node id in seperate rows of ids.
+                UNWIND {{node_id_list}} AS node_id
+
+                // for each node id in every row, collect the random list of its neighbors.
+                CALL apoc.cypher.run(
+
+                    'MATCH(cur_node) WHERE id(cur_node) = {{node_id}}
+
+                    // find the neighbors
+                    MATCH (cur_node){direction[sampling_direction]}(neighbors)
+
+                    // put all ids into a list
+                    WITH CASE collect(id(neighbors)) WHEN [] THEN [null] ELSE collect(id(neighbors)) END AS in_neighbors_list
+
+                    // pick random nodes with replacement
+                    WITH apoc.coll.randomItems(in_neighbors_list, {{num_samples}}, True) AS in_samples_list
+
+                    RETURN in_samples_list',
+                    {{ node_id: node_id, num_samples: {{num_samples}}  }}) YIELD value
+
+                RETURN apoc.coll.flatten(collect(value.in_samples_list)) as next_samples
+            """
+
+        self._check_neighbourhood_sizes(in_size, out_size)
+        self._check_common_parameters(nodes, n, len(in_size), seed)
+
+        samples = [[head_node for head_node in nodes for _ in range(n)]]
+
+        tree_cur_index = 0
+        tree_end_index = 0
+
+        in_sample_query = bfs_neighbor_query(sampling_direction="IN")
+        out_sample_query = bfs_neighbor_query(sampling_direction="OUT")
+
+        for in_num, out_num in zip(in_size, out_size):
+            # BFS traversal in the flattened binary tree
+            while tree_cur_index <= tree_end_index:
```

Comment:
This is a neat idea, but I think it's quite opaque and hard to understand. For one, it might be better to use `range`:

```python
previous_length = 0

for in_num, out_num in zip(in_size, out_size):
    initial_length = len(samples)
    # expand the neighbourhoods of the nodes that were added at the previous hop
    for i in range(previous_length, len(samples)):
        cur_nodes = samples[i]
        ...
    previous_length = initial_length
```

Alternatively, maybe we could model the hop structure more directly:

```python
# list of hops, which are a list of neighborhoods, which are a list of nodes
hops = [[nodes * n]]
for in_num, out_num in zip(in_size, out_size):
    last_hop = hops[-1]
    this_hop = []
    for neighbourhoods in last_hop:
        for cur_nodes in neighbourhoods:
             neighbor_records = ...
             this_hop.append(neighbor_records.data()[0]["next_samples"])
             ...

    hops.append(this_hop)

# flatten all of the hops to be left with just a list of all of the neighborhoods
return sum(hops)
```

Reply:
Sure! I am rewriting this part now.